### PR TITLE
Mesh shader: Change some mesh shader calls to dialects Op

### PR DESCRIPTION
**Pull request by amdrexu**
Define those new dialects Op:
- SetMeshOutputs
- SetMeshPrimitiveIndices
- SetMeshPrimitiveCulled
- GetMeshBuiltinInput
- WriteMeshVertexOutput
- WriteMeshPrimitiveOutput

Change the handling to dialects visitor and refactor their handlers to make them adapt to dialects visitor.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 61312c3794d2e0612001e45b9c2ffe4f8b5d0ab1
<details><summary>CTS tests (Failed: 258/211876)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 8.9, Gfx10</summary><ul>

<li>Passed: 35242/70631 (49.9%)</li>
<li><details><summary>Failed: 258/70631 (0.4%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.blit_image.all_formats.color.2d.e5b9g9r9_ufloat_pack32.a2r10g10b10_unorm_pack32.optimal_optimal_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.resolve_image.diff_image_size.dst_256_512_1_2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.blit_image.simple_tests.mirror_x_3d.r32_sfloat_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.depth_stencil_msaa_copy.whole.d32_sfloat_optimal_general_D_8_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.diff_image_size.src_256_512_1_8_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.external.semaphore.opaque_fd.export_import_signal_wait_permanent
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.2d.linear.remaining_array_layers.r16g16b16a16_unorm_33x128_clamp_input
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.1d.linear.remaining_array_layers_twostep.r16g16_sfloat
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.linear.remaining_array_layers_twostep.r16_snorm
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_depth_stencil_attachment.cube_layers.d16_unorm_s8_uint_separate_layouts_stencil
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_depth_stencil_attachment.multiple_layers.d32_sfloat_s8_uint_separate_layouts_stencil_1x33
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_depth_stencil_attachment.single_layer.d16_unorm_s8_uint_200x180
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.partial_clear_color_attachment.single_layer.b5g5r5a1_unorm_pack16_200x180
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.buffer_device_address.set31.depth2.baseubo.convertcheckuv2.store.single.std140.comp
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.buffer_device_address.set7.depth1.basessbo.convertcheckuv2.nostore.multi.std140.vert
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.descriptorset_random.sets16.unifindexed.ubolimitlow.nosbo.sampledimghigh.outimgtexlow.iublimithigh.nouab.frag.ialimitlow.0
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.mutable_descriptor.multiple.mixed.update_copy.nonmutable_source.normal_source.pool_no_types.update_after_bind.no_array.comp
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.mutable_descriptor.single.switches.storage_buffer_uniform_texel_buffer.update_copy.nonmutable_source.host_only_source.pool_same_types.update_after_bind.no_array.vert
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.mutable_descriptor.single.uniform_texel_buffer.update_copy.nonmutable_source.host_only_source.pool_same_types.pre_update.no_array.tese
Stack trace: Flake

FAILURE: dEQP-VK.binding_model.shader_access.primary_cmd_buf.with_template.combined_image_sampler_mutable.vertex.multiple_descriptor_sets.multiple_contiguous_descriptors.cube_array
Stack trace: Flake
...
```
</p></details></li>

<li>Not Supported: 35131/70631 (49.7%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39379/70629 (55.8%)</li>
<li>Failed: 0/70629 (0.0%)</li>
<li>Not Supported: 31250/70629 (44.2%)</li><li>Warnings: 0/70629 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70616 (55.8%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31203/70616 (44.2%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 61312c3794d2e0612001e45b9c2ffe4f8b5d0ab1
<details><summary>CTS tests (Failed: 0/211876)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35496/70631 (50.3%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35135/70631 (49.7%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39379/70629 (55.8%)</li>
<li>Failed: 0/70629 (0.0%)</li>
<li>Not Supported: 31250/70629 (44.2%)</li><li>Warnings: 0/70629 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70616 (55.8%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31203/70616 (44.2%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/interface/lgc/LgcDialect.td [hunk 1]
```diff
@@ -123,6 +123,88 @@ def EmitMeshTasksOp : LgcOp<"emit.mesh.tasks", [Memory<[]>]> {
...
+
+def SetMeshPrimitiveIndicesOp : LgcOp<"set.mesh.primitive.indices", [Memory<[]>]> {
+  let arguments = (ins I32:$primitiveIndex, (ScalarOrFixedVector I32):$primitiveIndices);
+  let results = (outs);
+
+  let summary = "set primitive indices for mesh shader";
+  let description = [{
+    In the mesh shader, set primitive indices by forming primitive connectivity data and writing it to LDS.
+
+    `primitiveIndex` is primitive indexing specifying which primitive to set.
```
Comment (nhaehnle) [inline lgc/interface/lgc/LgcDialect.td:148]: "is the primitve index"

File: lgc/interface/lgc/LgcDialect.td [hunk 2]
```diff
@@ -123,6 +123,88 @@ def EmitMeshTasksOp : LgcOp<"emit.mesh.tasks", [Memory<[]>]> {
...
+
+def SetMeshPrimitiveCulledOp : LgcOp<"set.mesh.primitive.culled", [Memory<[]>]> {
+  let arguments = (ins I32:$primitiveIndex, I1:$isCulled);
+  let results = (outs);
+
+  let summary = "set primitive culled state for mesh shader";
+  let description = [{
+    In the mesh shader, set primitive culled state by writing the null primitive flag to LDS.
+
+    `primitiveIndex` is primitive indexing specifying which primitive to set.
```
Comment (nhaehnle) [inline lgc/interface/lgc/LgcDialect.td:161]: as above

**Review by nhaehnle — COMMENTED** [review]
Thanks! A little nit inline, rest LGTM

File: lgc/interface/lgc/LgcDialect.td (hunk 1 above)
Comment (amdrexu) [reply lgc/interface/lgc/LgcDialect.td]: Done. Thanks.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9deffdf3910c88102204195de39acdb6ab53d417
<details><summary>CTS tests (Failed: 0/211876)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35496/70631 (50.3%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35135/70631 (49.7%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39379/70629 (55.8%)</li>
<li>Failed: 0/70629 (0.0%)</li>
<li>Not Supported: 31250/70629 (44.2%)</li><li>Warnings: 0/70629 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70616 (55.8%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31203/70616 (44.2%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
LGTM for the last commit, though you probably want to wait for the previous commits to have landed separately?

File: lgc/patch/VertexFetch.cpp [hunk 3]
```diff
@@ -121,11 +125,12 @@ class VertexFetchImpl : public VertexFetch {
...
 
-  LgcContext *m_lgcContext = nullptr;   // LGC context
-  LLVMContext *m_context = nullptr;     // LLVM context
-  Value *m_vertexBufTablePtr = nullptr; // Vertex buffer table pointer
-  Value *m_vertexIndex = nullptr;       // Vertex index
-  Value *m_instanceIndex = nullptr;     // Instance index
+  LgcContext *m_lgcContext = nullptr;      // LGC context
+  LLVMContext *m_context = nullptr;        // LLVM context
+  Value *m_vertexBufTablePtr = nullptr;    // Vertex buffer table pointer
+  Value *m_curAttribBufferDescr = nullptr; // Current attribute buffer descriptor;
```
Comment (tsymalla) [inline lgc/patch/VertexFetch.cpp:131]: Nit: superfluous ;

**Pull request merged**